src/lcm/grids.py

```diff
@@ -1,111 +1,259 @@
-"""Functions to generate and work with different kinds of grids.
+"""Collection of classes that are used by the user to define the model and grids."""
 
-Grid generation functions must have the following signature:
+import dataclasses as dc
+from abc import ABC, abstractmethod
+from collections.abc import Collection
+from dataclasses import dataclass, field
+from typing import NotRequired, TypedDict, cast
 
-    Signature (start: Scalar, stop: Scalar, n_points: int) -> jax.Array
+import jax.numpy as jnp
 
-They take start and end points and create a grid of points between them.
+from lcm.exceptions import GridInitializationError, format_messages
+from lcm.grid_helpers import linspace, logspace
+from lcm.interfaces import ContinuousGridInfo
+from lcm.typing import ContinuousGridType
 
+build_grid_mapping = {
+    "linspace": linspace,
+    "logspace": logspace,
+}
 
-Interpolation info functions must have the following signature:
 
-    Signature (
-        value: Scalar,
-        start: Scalar,
-        stop: Scalar,
-        n_points: int
-    ) -> Scalar
+class Grid(ABC):
+    """LCM Grid base class."""
 
-They take the information required to generate a grid, and return an index corresponding
-to the value, which is a point in the space but not necessarily a grid point.
+    @abstractmethod
+    def to_jax(self) -> jnp.ndarray:
+        """Convert the grid to a Jax array."""
 
-Some of the arguments will not be used by all functions but the aligned interface makes
-it easy to call functions interchangeably.
 
-"""
+@dataclass(frozen=True)
+class DiscreteGrid(Grid):
+    """A grid of discrete values.
+
+    Attributes:
+        options: The options in the grid. Must be an iterable of scalar int or float
+            values.
+
+    """
+
+    options: Collection[int | float]
+
+    def __post_init__(self) -> None:
+        if not isinstance(self.options, Collection):
+            raise GridInitializationError(
+                "options must be a collection of scalar int or float values, e.g., a ",
+                "list or tuple",
+            )
+
+        if errors := _validate_discrete_grid(self.options):
+            msg = format_messages(errors)
+            raise GridInitializationError(msg)
+
+    def to_jax(self) -> jnp.ndarray:
+        """Convert the grid to a Jax array."""
+        return jnp.array(list(self.options))
+
+    def replace(self, options: Collection[int | float]) -> "DiscreteGrid":
```

Comment:
Why is the `replace` method ever needed? You can just as easily create a new one.